src/lib/HST/Frontend/FromGHC.hs

```diff
@@ -0,0 +1,382 @@
+  S.PTuple (transformSrcSpan s) (transformBoxity boxity) (map transformPat pats)
+transformPat (GHC.L s (GHC.ParPat _ pat)) =
+  S.PParen (transformSrcSpan s) (transformPat pat)
+transformPat (GHC.L s (GHC.ListPat _ pats)) =
+  S.PList (transformSrcSpan s) (map transformPat pats)
+transformPat (GHC.L s (GHC.WildPat _)) = S.PWildCard (transformSrcSpan s)
+transformPat _                         = error "Unsupported pattern"
+
+-- | Transforms a GHC module name with an HST source span into an HST module
+--   name.
+transformModuleName :: S.SrcSpan GHC -> GHC.ModuleName -> S.ModuleName GHC
+transformModuleName s modName = S.ModuleName s (GHC.moduleNameString modName)
+
+-- | Transforms a GHC located reader name into an HST qualified name and a
+--   @Bool@ which is @True@ if the name belongs to a data constructor and
+--   @False@ otherwise.
+transformRdrName :: GHC.Located GHC.RdrName -> (S.QName GHC, Bool)
+transformRdrName (GHC.L s (GHC.Unqual name)) =
+  let s' = transformSrcSpan s
+  in  (S.UnQual s' (S.Ident s' (GHC.occNameString name)), GHC.isDataOcc name)
+transformRdrName (GHC.L s (GHC.Qual modName name)) =
+  let s' = transformSrcSpan s
+  in  ( S.Qual s'
+               (transformModuleName s' modName)
+               (S.Ident s' (GHC.occNameString name))
+      , GHC.isDataOcc name
+      )
+transformRdrName (GHC.L s (GHC.Exact name)) =
+  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), False)
+transformRdrName _ = error "Unsupported RdrName"
+
+-- | Transforms a GHC located unqualified reader name into an HST name.
+transformRdrNameUnqual :: GHC.Located GHC.RdrName -> S.Name GHC
+transformRdrNameUnqual (GHC.L s (GHC.Unqual occName)) =
+  S.Ident (transformSrcSpan s) (GHC.occNameString occName)
+transformRdrNameUnqual _ = error "Expected an unqualified name"
+
+-- | Transforms a GHC name with an HST source span into an HST special
+--   constructor.
+transformName :: S.SrcSpan GHC -> GHC.Name -> S.SpecialCon GHC
+transformName s name = case lookup name (specialConMap s) of
+  Just sCon -> sCon
+  Nothing   -> case GHC.wiredInNameTyThing_maybe name of
+    -- TODO Can the other TyThing constructors occur?
+    Just (GHC.ATyCon tc) -> case GHC.tyConTuple_maybe tc of
+      Just GHC.BoxedTuple   -> S.TupleCon s S.Boxed (GHC.tyConArity tc)
+      Just GHC.UnboxedTuple -> S.TupleCon s S.Unboxed (GHC.tyConArity tc)
+      _                     -> error "Unexpected special constructor"
+    _ -> error "Expected a built-in special constructor"
+
+-- | Takes an HST source span and maps GHC names to HST special constructors.
+specialConMap :: S.SrcSpan GHC -> [(GHC.Name, S.SpecialCon GHC)]
+specialConMap s =
+  [ (GHC.tyConName GHC.unitTyCon          , S.UnitCon s)
+  , (GHC.listTyConName                    , S.ListCon s)
+  , (GHC.tyConName GHC.funTyCon           , S.FunCon s)
+  , (GHC.consDataConName                  , S.Cons s)
+  , (GHC.tupleTyConName GHC.UnboxedTuple 1, S.UnboxedSingleCon s)
+  ]
```

Comment:
That the special case for `S.TupleCon` is not covered by `specialConMap` but is implemented in `transformName` should at least be documented. However, for error reporting purposes it would be better if all names were transformed in `transformName` using pattern matching.

Reply:
I wanted to use pattern matching, but the [`Name`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/Name.html#v:Name) and [`OccName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/OccName.html#t:OccName) types of `ghc-lib-parser` don't export their constructors, which is why I chose this solution.
I will mention the special case for tuple constructors in the documentation, though.